// profile-pane/lib/editProfilePane/editProfile.view.js
"use strict";
/**
 * Profile Editing App Pane
 *
 * Unlike view panes, this is available any place whatever the real subject,
 * and allows the user to edit their own profile.
 *
 * Usage: paneRegistry.register('profile/profilePane')
 * or standalone script adding onto existing mashlib.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var rdflib_1 = require("rdflib");
var solid_ui_1 = require("solid-ui");
var profile_dom_1 = require("./profile.dom");
var wrapped_profileFormText_1 = require("./wrapped-profileFormText");
var highlightColor = solid_ui_1.style.highlightColor || '#7C4DFF';
var editProfileView = {
    global: true,
    icon: solid_ui_1.icons.iconBase + 'noun_492246.svg',
    name: 'editProfile',
    label: function () { return null; }, // don't use this in the normal solid-panes dispatching
    render: function (subject, context) {
        console.log('@@@ render edit profile pane:  subject, context', subject, context);
        var dom = context.dom;
        var store = context.session.store;
        function complainIfBad(ok, mess) {
            if (ok)
                return;
            div.appendChild(solid_ui_1.widgets.errorMessageBlock(dom, mess, '#fee'));
        }
        function renderProfileForm(div, subject) {
            var preferencesForm = (0, rdflib_1.sym)('https://solidos.github.io/profile-pane/src/ontology/profileForm.ttl#this');
            var preferencesFormDoc = preferencesForm.doc();
            if (!store.holds(undefined, undefined, undefined, preferencesFormDoc)) {
                // If not loaded already
                (0, rdflib_1.parse)(wrapped_profileFormText_1.profileForm, store, preferencesFormDoc.uri, 'text/turtle', function () { return null; }); // Load form directly
            }
            div.setAttribute('data-testid', 'profile-editor');
            // @@ div.append?
            solid_ui_1.widgets.appendForm(dom, div, {}, subject, preferencesForm, editableProfile, complainIfBad);
        } // renderProfileForm
        var div = dom.createElement('div');
        div.setAttribute('data-testid', 'profile-editor');
        var editableProfile;
        div.setAttribute('style', "border: 0.3em solid ".concat(highlightColor, "; border-radius: 0.5em; padding: 0.7em; margin-top:0.7em;"));
        var table = div.appendChild(dom.createElement('table'));
        // const top = table.appendChild(dom.createElement('tr'))
        var main = table.appendChild(dom.createElement('tr'));
        var bottom = table.appendChild(dom.createElement('tr'));
        var statusArea = bottom.appendChild(dom.createElement('div'));
        statusArea.setAttribute('style', 'padding: 0.7em;');
        function comment(str) {
            var p = main.appendChild(dom.createElement('p'));
            p.setAttribute('style', 'padding: 1em;');
            p.textContent = str;
            return p;
        }
        function heading(str) {
            var h = main.appendChild(dom.createElement('h3'));
            h.setAttribute('style', 'color:' + highlightColor + ';');
            h.textContent = str;
            return h;
        }
        var profileContext = {
            dom: dom,
            div: main,
            statusArea: statusArea,
            me: null
        };
        solid_ui_1.login.ensureLoadedProfile(profileContext)
            .then(function (theContext) {
            var me = theContext.me;
            heading('Edit your public profile');
            var profile = me.doc();
            if (!store.updater) {
                throw new Error('Store has no updater');
            }
            if (store.any(me, solid_ui_1.ns.solid('editableProfile'))) {
                editableProfile = store.any(me, solid_ui_1.ns.solid('editableProfile'));
            }
            else if (store.updater.editable(profile.uri, store)) {
                editableProfile = profile;
            }
            else {
                statusArea.appendChild(solid_ui_1.widgets.errorMessageBlock(dom, "\u26A0\uFE0F Your profile ".concat(profile, " is not editable, so we cannot do much here."), 'straw'));
                return;
            }
            comment("Everything you put here will be public.\n        There will be other places to record private things.");
            heading('Your contact information');
            main.appendChild((0, profile_dom_1.paneDiv)(context, me, 'contact'));
            heading('People you know who have WebIDs');
            comment("This is your public social network.\n        Only put people here to whom you are happy to be publicly connected.\n        (You can always keep private track of friends and family in your contacts.)");
            // TODO: would be useful to explain what it means to "drag people"
            //       what is it that is being dragged?
            //       is there a way to search for people (or things to drag) on this page?
            if (editableProfile) {
                comment('Drag people onto the target below to add people.');
            }
            solid_ui_1.widgets.attachmentList(dom, me, main, {
                doc: profile,
                modify: !!editableProfile,
                predicate: solid_ui_1.ns.foaf('knows'),
                noun: 'friend'
            });
            heading('Communities you participate in');
            comment("These are organizations and projects (etc) whose stuff you share");
            // TODO: would be useful to explain what it means to "drag organizations"
            //       what is it that is being dragged?
            //       is there a way to search for people (or things to drag) on this page?
            //  Also provide a way of using cut and paste
            if (editableProfile) {
                comment('Drag organizations onto the target below to add organizations.');
            }
            solid_ui_1.widgets.attachmentList(dom, me, main, {
                doc: profile,
                modify: !!editableProfile,
                predicate: solid_ui_1.ns.solid('community'),
                noun: 'community'
            });
            renderProfileForm(main, me);
            heading('Thank you for filling your profile.');
        }).catch(function (error) {
            statusArea.appendChild(solid_ui_1.widgets.errorMessageBlock(dom, error, '#fee'));
        });
        return div;
    }
};
exports.default = editProfileView;
