// profile-pane/lib/integration-tests/helpers/dataSetup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ClubPublicTypes = exports.ClubPublicTypeIndex = exports.ClubProfile = exports.ClubPrivateTypes = exports.ClubPrivateTypeIndex = exports.ClubPreferencesFile = exports.ClubPreferences = exports.BobProfile = exports.AlicePublicTypes = exports.AlicePublicTypeIndex = exports.AliceProfileFile = exports.AliceProfile = exports.AlicePrivateTypes = exports.AlicePrivateTypeIndex = exports.AlicePreferencesFile = exports.AlicePhotos = exports.AlicePreferences = exports.AlicePhotoFolder = exports.club = exports.bobby = exports.bob = exports.alice = void 0;
exports.loadWebObject = loadWebObject;
exports.clearLocalStore = clearLocalStore;
var rdflib_1 = require("rdflib");
//------ Club -------------------------------------------------------
var club = (0, rdflib_1.sym)('https://club.example.com/profile/card.ttl#it');
exports.club = club;
var ClubPreferencesFile = (0, rdflib_1.sym)('https://club.example.com/settings/prefs.ttl');
exports.ClubPreferencesFile = ClubPreferencesFile;
var ClubPublicTypeIndex = (0, rdflib_1.sym)('https://club.example.com/profile/public-type-index.ttl');
exports.ClubPublicTypeIndex = ClubPublicTypeIndex;
var ClubPrivateTypeIndex = (0, rdflib_1.sym)('https://club.example.com/settings/private-type-index.ttl');
exports.ClubPrivateTypeIndex = ClubPrivateTypeIndex;
var ClubProfile = "\n\n<#it> a vcard:Organization;\n    space:preferencesFile ".concat(ClubPreferencesFile, ";\n    solid:publicTypeIndex ").concat(ClubPublicTypeIndex, ";\n    vcard:fn \"Card Club\" .\n");
exports.ClubProfile = ClubProfile;
var ClubPreferences = "\n    ".concat(club, " solid:privateTypeIndex ").concat(ClubPrivateTypeIndex, " .\n");
exports.ClubPreferences = ClubPreferences;
var ClubPublicTypes = "\n\n:chat1 solid:forClass meeting:LongChat; solid:instance <../publicStuff/ourChat.ttl#this> .\n\n:todo solid:forClass wf:Tracker; solid:instance  <../publicStuff/actionItems.ttl#this>.\n\n:issues solid:forClass wf:Tracker; solid:instance  <../project4/clubIssues.ttl#this>.\n";
exports.ClubPublicTypes = ClubPublicTypes;
var ClubPrivateTypes = "\n:id1592319218311 solid:forClass wf:Tracker; solid:instance  <../privateStuff/ToDo.ttl#this>.\n\n:id1592319391415 solid:forClass wf:Tracker; solid:instance <../privateStuff/Goals.ttl#this>.\n\n:id1595595377864 solid:forClass wf:Tracker; solid:instance <../privateStuff/tasks.ttl#this>.\n\n:id1596123375929 solid:forClass meeting:Meeting; solid:instance  <../project4/clubMeeting.ttl#this>.\n\n";
exports.ClubPrivateTypes = ClubPrivateTypes;
//------ Alice -------------------------------------------------------
var alice = (0, rdflib_1.sym)("https://alice.example.com/profile/card.ttl#me");
exports.alice = alice;
var AliceProfileFile = alice.doc();
exports.AliceProfileFile = AliceProfileFile;
var AlicePreferencesFile = (0, rdflib_1.sym)('https://alice.example.com/settings/prefs.ttl');
exports.AlicePreferencesFile = AlicePreferencesFile;
var AlicePublicTypeIndex = (0, rdflib_1.sym)('https://alice.example.com/profile/public-type-index.ttl');
exports.AlicePublicTypeIndex = AlicePublicTypeIndex;
var AlicePrivateTypeIndex = (0, rdflib_1.sym)('https://alice.example.com/settings/private-type-index.ttl');
exports.AlicePrivateTypeIndex = AlicePrivateTypeIndex;
var AlicePhotoFolder = (0, rdflib_1.sym)(alice.dir().uri + "Photos/");
exports.AlicePhotoFolder = AlicePhotoFolder;
var AliceProfile = "\n<#me> a vcard:Individual;\n    space:preferencesFile ".concat(AlicePreferencesFile, ";\n    solid:publicTypeIndex ").concat(AlicePublicTypeIndex, ";\n    vcard:fn \"Alice\" .\n");
exports.AliceProfile = AliceProfile;
var AlicePreferences = "\n    ".concat(alice, " solid:privateTypeIndex ").concat(AlicePrivateTypeIndex, ";\n    solid:community ").concat(club, " .\n");
exports.AlicePreferences = AlicePreferences;
var AlicePublicTypes = "\n\n:chat1 solid:forClass meeting:LongChat; solid:instance <../publicStuff/myChat.ttl#this> .\n\n:todo solid:forClass wf:Tracker; solid:instance  <../publicStuff/actionItems.ttl#this>.\n\n:issues solid:forClass wf:Tracker; solid:instance  <../project4/issues.ttl#this>.\n\n:photos solid:forClass schema:Image; solid:instanceContainer  ".concat(AlicePhotoFolder, " .\n");
exports.AlicePublicTypes = AlicePublicTypes;
var AlicePrivateTypes = "\n:id1592319218311 solid:forClass wf:Tracker; solid:instance  <../privateStuff/ToDo.ttl#this>.\n\n:id1592319391415 solid:forClass wf:Tracker; solid:instance <../privateStuff/Goals.ttl#this>.\n\n:id1595595377864 solid:forClass wf:Tracker; solid:instance <../privateStuff/workingOn.ttl#this>.\n\n:id1596123375929 solid:forClass meeting:Meeting; solid:instance  <../project4/meeting1.ttl#this>.\n\n";
exports.AlicePrivateTypes = AlicePrivateTypes;
var AlicePhotos = "\n<>\n    a ldp:BasicContainer, ldp:Container;\n    dct:modified \"2021-04-26T05:34:16Z\"^^xsd:dateTime;\n    ldp:contains\n        <photo1.png>, <photo2.png>, <photo3.png> ;\n    stat:mtime 1619415256.541;\n    stat:size 4096 .\n";
exports.AlicePhotos = AlicePhotos;
//------ Bob -------------------------------------------------------
var bob = (0, rdflib_1.sym)('https://bob.example.com/profile/card.ttl#me');
exports.bob = bob;
var BobProfile = "\n<#me> a vcard:Individual;\nvcard:fn \"Bob\" .\n";
exports.BobProfile = BobProfile;
//------ Bobby -------------------------------------------------------
var bobby = (0, rdflib_1.sym)('https://bobby.example.com/profile/card.ttl#me');
exports.bobby = bobby;
var BobbyProfile = "\n<#me> a vcard:Individual;\nvcard:fn \"Bobby\" .\n";
function loadWebObject() {
    var web = {};
    web[alice.doc().uri] = AliceProfile;
    web[AlicePreferencesFile.uri] = AlicePreferences;
    web[AlicePrivateTypeIndex.uri] = AlicePrivateTypes;
    web[AlicePublicTypeIndex.uri] = AlicePublicTypes;
    web[AlicePhotoFolder.uri] = AlicePhotos;
    web[bob.doc().uri] = BobProfile;
    web[bobby.doc().uri] = BobbyProfile;
    web[club.doc().uri] = ClubProfile;
    web[ClubPreferencesFile.uri] = ClubPreferences;
    web[ClubPrivateTypeIndex.uri] = ClubPrivateTypes;
    web[ClubPublicTypeIndex.uri] = ClubPublicTypes;
    return web;
}
function clearLocalStore(store) {
    store.statements.slice().forEach(store.remove.bind(store));
}
