// profile-pane/lib/0SAVED-CVCard.js
"use strict";
var __makeTemplateObject = (this && this.__makeTemplateObject) || function (cooked, raw) {
    if (Object.defineProperty) { Object.defineProperty(cooked, "raw", { value: raw }); } else { cooked.raw = raw; }
    return cooked;
};
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CVCard = void 0;
var lit_html_1 = require("lit-html");
var baseStyles_1 = require("./baseStyles");
var style_map_js_1 = require("lit-html/directives/style-map.js");
var baseStyles_2 = require("./baseStyles");
var styles = {
    image: (0, style_map_js_1.styleMap)((0, baseStyles_1.fullWidth)()),
    intro: (0, style_map_js_1.styleMap)(__assign(__assign({}, (0, baseStyles_1.textGray)()), (0, baseStyles_1.textCenter)())),
    card: (0, style_map_js_1.styleMap)((0, baseStyles_2.card)()),
    info: (0, style_map_js_1.styleMap)(__assign(__assign({}, (0, baseStyles_1.paddingSmall)()), (0, baseStyles_1.textLeft)())),
};
var CVCard = function (profileBasics, cvData) {
    var rolesByType = cvData.rolesByType, skills = cvData.skills, languages = cvData.languages;
    var nameStyle = (0, style_map_js_1.styleMap)(__assign(__assign({}, (0, baseStyles_1.heading)()), { 
        // "text-decoration": "underline",
        color: profileBasics.highlightColor }));
    if (renderRoles(rolesByType["FutureRole"]) || renderRoles(rolesByType["CurrentRole"]) || renderRoles(rolesByType["PastRole"]) || renderSkills(skills) || renderLanguages(languages)) {
        return (0, lit_html_1.html)(templateObject_1 || (templateObject_1 = __makeTemplateObject(["\n  <div data-testid=\"curriculum-vitae\" style=\"", "\">\n    <div style=", ">\n      <h3 style=", ">Bio</h3>\n      <div style=", ">", "</div>\n      <hr />\n      <div style=", ">", "</div>\n      <hr />\n      <div style=", ">", "</div>\n      <hr />\n      <h3 style=", ">Skills</h3>\n      <div style=", ">", "</div>\n      <h3 style=", ">Languages</h3>\n      <div style=", ">", "</div>\n    </div>\n    </div>\n  "], ["\n  <div data-testid=\"curriculum-vitae\" style=\"", "\">\n    <div style=", ">\n      <h3 style=", ">Bio</h3>\n      <div style=", ">", "</div>\n      <hr />\n      <div style=", ">", "</div>\n      <hr />\n      <div style=", ">", "</div>\n      <hr />\n      <h3 style=", ">Skills</h3>\n      <div style=", ">", "</div>\n      <h3 style=", ">Languages</h3>\n      <div style=", ">", "</div>\n    </div>\n    </div>\n  "])), styles.card, styles.info, nameStyle, styles.info, renderRoles(rolesByType["FutureRole"]), styles.info, renderRoles(rolesByType["CurrentRole"]), styles.info, renderRoles(rolesByType["PastRole"]), nameStyle, styles.info, renderSkills(skills), nameStyle, styles.info, renderLanguages(languages));
    }
    return (0, lit_html_1.html)(templateObject_2 || (templateObject_2 = __makeTemplateObject([""], [""])));
};
exports.CVCard = CVCard;
function renderRole(role) {
    return role
        ? (0, lit_html_1.html)(templateObject_3 || (templateObject_3 = __makeTemplateObject(["<div style=\"margin-top: 0.3em; margin-bottom: 0.3em;\">\n        <b>", "</b>\n        <span>", "</span>\n        <span>", "</span>   \n      </div> "], ["<div style=\"margin-top: 0.3em; margin-bottom: 0.3em;\">\n        <b>", "</b>\n        <span>", "</span>\n        <span>", "</span>   \n      </div> "])), role.orgName, strToUpperCase(role.roleText), role.dates) : (0, lit_html_1.html)(templateObject_4 || (templateObject_4 = __makeTemplateObject([""], [""])));
}
function renderRoles(roles) {
    if (roles[0] > "")
        return (0, lit_html_1.html)(templateObject_6 || (templateObject_6 = __makeTemplateObject(["", "", ""], ["", "", ""])), renderRole(roles[0]), roles.length > 1 ? renderRoles(roles.slice(1)) : (0, lit_html_1.html)(templateObject_5 || (templateObject_5 = __makeTemplateObject([""], [""]))));
}
function renderSkill(skill) {
    return skill
        ? (0, lit_html_1.html)(templateObject_7 || (templateObject_7 = __makeTemplateObject(["<div style=\"margin: 0.5em;\">\n        <p style=\"text-align: center;\">", "</p>\n      </div> "], ["<div style=\"margin: 0.5em;\">\n        <p style=\"text-align: center;\">", "</p>\n      </div> "])), skill) : (0, lit_html_1.html)(templateObject_8 || (templateObject_8 = __makeTemplateObject([""], [""])));
}
function renderSkills(skills) {
    if (skills[0] > "")
        return (0, lit_html_1.html)(templateObject_10 || (templateObject_10 = __makeTemplateObject(["", " ", ""], ["", " ", ""])), renderSkill(strToUpperCase(skills[0])), skills.length > 1 ? renderSkills(skills.slice(1)) : (0, lit_html_1.html)(templateObject_9 || (templateObject_9 = __makeTemplateObject([""], [""]))));
}
function renderLan(language) {
    return language
        ? (0, lit_html_1.html)(templateObject_11 || (templateObject_11 = __makeTemplateObject(["<div style=\"margin: 0.5em;\">\n        <p style=\"text-align: center;\">", "</p>\n      </div> "], ["<div style=\"margin: 0.5em;\">\n        <p style=\"text-align: center;\">", "</p>\n      </div> "])), language) : (0, lit_html_1.html)(templateObject_12 || (templateObject_12 = __makeTemplateObject([""], [""])));
}
function renderLanguages(languages) {
    if (languages[0] > "")
        return (0, lit_html_1.html)(templateObject_14 || (templateObject_14 = __makeTemplateObject(["", "", ""], ["", "", ""])), renderLan(languages[0]), languages.length > 1 ? renderLanguages(languages.slice(1)) : (0, lit_html_1.html)(templateObject_13 || (templateObject_13 = __makeTemplateObject([""], [""]))));
}
function strToUpperCase(str) {
    if (str && str[0] > "") {
        var strCase = str.split(' ');
        for (var i = 0; i < strCase.length; i++) {
            strCase[i] = strCase[i].charAt(0).toUpperCase() +
                strCase[i].substring(1);
        }
        return strCase.join(' ');
    }
    return '';
}
var templateObject_1, templateObject_2, templateObject_3, templateObject_4, templateObject_5, templateObject_6, templateObject_7, templateObject_8, templateObject_9, templateObject_10, templateObject_11, templateObject_12, templateObject_13, templateObject_14;
