// profile-pane/lib/integration-tests/setup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.context = exports.doc = exports.subject = void 0;
exports.fakeLogInAs = fakeLogInAs;
exports.delay = delay;
var rdflib_1 = require("rdflib");
var solid_logic_1 = require("solid-logic");
exports.subject = (0, rdflib_1.sym)("https://janedoe.example/profile/card#me");
exports.doc = exports.subject.doc();
/*
if (this && this.session && this.session.info && this.session.info.webId && this.session.info.isLoggedIn)
{ return (0, rdflib_1.sym)(this.session.info.webId);

*/
function fakeLogInAs(subject) {
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    window.$SolidTestEnvironment = {
        username: subject ? subject.value : null // assume logged in if not null
    };
}
function delay(ms) {
    return new Promise(function (resolve) { return setTimeout(resolve, ms); });
}
exports.context = {
    dom: document,
    getOutliner: function () { return null; },
    session: {
        info: {
            isLoggedIn: true, // Added for editProofilepane
            webId: exports.subject.value,
        },
        paneRegistry: {
            byName: function (name) {
                return {
                    render: function () {
                        return document.createElement('div')
                            .appendChild(document.createTextNode("mock ".concat(name, " pane")));
                    }
                };
            }
        },
        store: solid_logic_1.store,
        logic: {},
    },
};
